### PR TITLE
Manage Packages: optimize package link generation

### PR DESCRIPTION
**Pull request by chenriksson**
Seeing up to 60% increase in server-side view performance. See #5120 for details.

### CODE REVIEW

File: src/NuGetGallery/App_Code/ViewHelpers.cshtml [hunk 1]
```diff
@@ -246,6 +246,17 @@
     </span>
 }
 
+@helper UserLinkWithIcon(RouteValueTemplate<User> urlTemplate, User user)
```
Comment (chenriksson) [inline src/NuGetGallery/App_Code/ViewHelpers.cshtml]: Was used before PR #5060 to show org icon next to org owners. Will fix in morning to use again.

File: src/NuGetGallery/Helpers/RouteValueTemplate.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Web;
+using System.Web.Routing;
+
+namespace NuGetGallery.Helpers
+{
+    public class RouteValueTemplate<T>
```
Comment (chenriksson) [inline src/NuGetGallery/Helpers/RouteValueTemplate.cs]: Maybe RouteValuesTemplate is better name? Or ActionLinkTemplate?

File: src/NuGetGallery/UrlExtensions.cs [hunk 3]
```diff
@@ -3,12 +3,13 @@
 
 using NuGetGallery.Configuration;
```
Comment (chenriksson) [inline src/NuGetGallery/UrlExtensions.cs]: nit: will sort

File: src/NuGetGallery/Helpers/RouteValueTemplate.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        public RouteValueTemplate(Func<RouteValueDictionary, string> linkGenerator, IDictionary<string, Func<T, object>> routesGenerator)
+        {
+            RoutesGenerator = routesGenerator;
+
+            var templateValues = GetTemplateValues();
+            var encodedLinkTemplate = linkGenerator(templateValues);
+            LinkTemplate = HttpUtility.UrlDecode(encodedLinkTemplate);
+        }
+
+        private string LinkTemplate { get; set; }
```
Comment (skofman1) [inline src/NuGetGallery/Helpers/RouteValueTemplate.cs]: nit: the private properties can be made fields.

File: src/NuGetGallery/App_Code/ViewHelpers.cshtml (hunk 1 above)
Comment (chenriksson) [reply src/NuGetGallery/App_Code/ViewHelpers.cshtml]: reverted, will do separately

File: src/NuGetGallery/Helpers/RouteValueTemplate.cs (hunk 2 above)
Comment (chenriksson) [reply src/NuGetGallery/Helpers/RouteValueTemplate.cs]: Renamed to RouteUrlTemplate

File: src/NuGetGallery/Helpers/RouteValueTemplate.cs (hunk 4 above)
Comment (chenriksson) [reply src/NuGetGallery/Helpers/RouteValueTemplate.cs]: done

File: src/NuGetGallery/UrlExtensions.cs (hunk 3 above)
Comment (chenriksson) [reply src/NuGetGallery/UrlExtensions.cs]: done

File: src/NuGetGallery/UrlExtensions.cs [hunk 5]
```diff
@@ -213,6 +214,25 @@ public static string PackageList(this UrlHelper url, bool relativeUrl = true)
...
+            this UrlHelper url,
+            bool relativeUrl = true)
+        {
+            var routesGenerator = new Dictionary<string, Func<IPackageVersionModel, object>>
+            {
+                { "id", p => p.Id },
+                { "version", p => null }
+            };
+
+            Func<RouteValueDictionary, string> linkGenerator = rv => GetRouteLink(
```
[the remark this answers is not in the document]
Comment (chenriksson) [reply src/NuGetGallery/UrlExtensions.cs]: Added comments to clarify when to use `Url.PackageRegistrationTemplate` and when to use `Url.Package`.

For code reusability, there's a few ways to go here. Like you said, I could create a shared helper... or I could just always use the template to create the package links. The templated version reduces expensive GetVirtualPath calls, but has an overhead of more allocations. Right now I only use the template per-request, but I could cache it so it's used across requests and also by the Url.Package, etc APIs.

Opened issue #5162 to track... just want to think about it more, and do in separate PR.

**Pull request merged**